manpages/xgobi.1
.TH XGobi 1 "30 September 1997"
.SH NAME
.B xgobi 
\- interactive dynamic graphics program for data visualization
.SH SYNOPSIS
.B xgobi
[ X options ]
[
.B \-subset
n
] [
.B \-only
n/N or a,n
] [
.B \-mono
] [
.B \-version
] [
.B \-std
mmx|msd|mmd
] [
.B \-scatmat
]
filename
.SH DESCRIPTION
.I xgobi
is an interactive dynamic graphics program for data
visualization in the X Window System.  It is especially
designed for the exploration of multivariate data.
Its basic plot is a scatterplot, and these are some
of the tools available for scatterplot display and
manipulation:
.RS .25i
.IP o 2
Cycling rapidly through two-variable scatter plots.
.IP o 2
Three-dimensional rotation, including trackball control, in spaces
spanned by three variables at a time.
.IP o 2
Grand tours and correlation tours: smooth randomized sequences of
two-dimensional projections in order to explore a higher-dimensional
point cloud of multivariate data.  Projections can be manually
controlled and optimized with projection pursuit.
.IP o 2
Brushing: the ability to change the color or plotting
character of a point or a group of points.
.IP o 2
Hiding groups:  using checkboxes to specify which brushed
groups should be plotted, or included in calculations.
.IP o 2
Scaling: moving and reshaping the plot interactively.
.IP o 2
Identification: displaying a label next to a point in the plot window.
.IP o 2
Linked views: Brushing, identification and touring are linked; that
is, actions in the window of one XGobi process are immediately
reflected in another XGobi window displaying the same data.
.IP o 2
Line editing: Any pair of points can be connected with a line segment.
.IP o 2
Moving points:  Points can be moved on the screen.  In
higher-dimensional views, the motion is interpreted as
taking place in some plane parallel to the screen.
.IP o 2
Smooths: Smooth curves can be drawn through point clouds.
.IP o 2
Subsetting: Subsamples can be drawn either systematically or randomly.
.IP o 2
Jittering: Variables can be jittered, which is useful for discrete
variables that take on only few values.
.IP o 2
Parallel coordinate display: shown in a separate linked window,
linked to the main window.
.IP o 2
Case label list: shown in a separate window, linked to the main
window for labeling points.
.IP o 2
Missing values: are accepted and can be dealt with by imputation of
constant values, random values, or user-supplied imputed values.
Missing value patterns can be examined in a separate linked XGobi
window.
.IP o 2
Variable transformations: A menu of transformations such as logs and a
few powers is available.
.IP o 2
Postscript plotting: The contents of the plot windows can be written
out to postscript files for high-quality plotting.
.IP o 2
Online help: Most of the XGobi buttons and windows have associated
help files that can be displayed during an XGobi session.  Click on
the
.I Info
button toward the top right of the window for
instructions.
.RE

XGobi has a direct manipulation interface, and all the above actions
are performed using the mouse.

XGobi can be used in conjunction with the S language for
scientific computing and data analysis.  Execute
.IR help(xgobi) 
within S or Splus for information; if the xgobi function
has not been installed, investigate the XGobi distribution files.

XGobi can also be programmed to use RPC (Remote Procedure Calls)
to communicate with other software.

.SH OPTIONS

.B \-subset
n
.LP
.RS .25i
Specify the size n of the random sample of the data to be displayed on
startup.  Although only the random sample is shown, all the data is
read in and available during the session.  Use the subset panel in
Tools to change the sample size and sampling method during the
session.
.RE

.B \-only
n/N
.LP
.RS .25i
Specify the size n of the random sample of the data to be read in on
startup from a file of N rows.  In constrast to subset, this option does
.B not
read in all data.
.RE

.B \-only
a,n
.LP
.RS .25i
Specify the initial row a and the number of rows n of the consecutive block
in data to read in.  In constrast to subset, this option does
.B not
read in all data.
.RE

.B \-std
mmx|msd|mmd
.LP
.RS .25i
By default, the data are scaled into the plotting window using the
minimum and maximum values of each variable or variable group, in such
a way that the midpoint of the variable is at the center of the
plotting window and no points fall outside the window.  Instead, to
scale using mean and largest distance, specify
.IR \-std
.IR msd;
to scale using the median and largest distance,
specify
.IR \-std
.IR mmd.
.RE

.B \-dev
x
.LP
.RS .25i
If you have specified 
.IR -std
.IR msd
or
.IR -std
.IR mmd,
then you can also specify the number of standard deviations (or median
absolute deviations) from the mean (or median) to be contained within
the plotting window, using the argument
.IR -dev
.IR x,
where
.IR x
is a real number between 0 and 100.  The default is 2.
.RE

.B \-mono
.LP
.RS .25i
Emulate a black-and-white display.  This simplifies producing pictures
for many publications.
.RE

.B \-version
.LP
.RS .25i
Print a version date, the last date that any file in the release
was changed.
.RE

.B \-scatmat
.LP
.RS .25i
Build a scatterplot matrix from the input file, and display it
in the plotting window.  This matrix is linked by default to
non-scatmat xgobi windows.
.RE

X options

.RS .25i
The standard X command line options can be used with XGobi.  These
include
.IR "-display machinename:0",
used when running an X program on one machine and displaying its
output on another, and
.IR "-title Title",
where
.IR Title
is a string you want to appear in the window manager titlebar.
.RE
.SH FILES
XGobi accepts standard input, but is most often used with files,
partly because of the additional plot control that can be achieved
using a set of files.  The data input file should be an ASCII file
with the data matrix arranged in rows and columns; in ASCII, rows must
be distinguished by carriage returns, and columns can be separated by
any amount of white space.  Missing values can be coded as ".", "NA"
or "na".  (The input file can also be a binary file, which
can be produced within XGobi once the ASCII data has been read in.)
XGobi accepts other input about the display of the data from files as
well.  If the data is in a file named

.nf
.B filename
.fi

.RS .25i
or
.RE

.nf
.B filename.dat
.fi

.RS .25i
(either of which must be an ASCII file), or
.RE

.nf
.B filename.bin
.fi
.RS .25i

(the binary version of the data), then the other files are as follows:
.RE

.nf
.B filename.row
.B filename.rowlab
.B filename.case
.fi
.RS .25i

Row or case labels: a label for each row of the data matrix, which is
displayed in the identification mode.  The file should contain one
label per line.

.RE

.nf
.B filename.col
.B filename.collab
.B filename.column
.B filename.var
.fi
.RS .25i

Column or variable labels: a label for each column of the data matrix,
which becomes part of the XGobi variable selection panel.  The file
should contain one label per line. 

To add a short label to be used in the parallel coordinates plot, use
the vertical bar "|" as a field separator and add the short name, so
each line looks like this:  fullname|shortname.
.RE

.nf
.B filename.colors
.fi
.RS .25i

Brushing colors: a color for each point in the plot, representing a
row or case of the data.  The file should contain one color per line.
(It is probably best if the colors correspond to the colors used in
brushing; see the later section on resources.)
.RE

.nf
.B filename.glyphs
.fi
.RS .25i

Brushing glyphs: a glyph type for each point in the plot, representing
a row or case of the data.  The file should contain one glyph type per
line.  The glyph types are as follows:
.RS .25i

.nf
1 through 5: Five sizes of '+'
6 through 10: Five sizes of 'X'
11 through 15: Five sizes of open rectangle
16 through 20: Five sizes of filled rectangle
21 through 25: Five sizes of open circle
26 through 30: Five sizes of filled circle
31: A single-pixel point
.fi
.RE

.RE
.nf
.B filename.erase
.fi
.RS .25i

Erase: a column of 1s (to have a point hidden on startup) and 0s (to
have the point plotted).  There should be one value per line and as
many lines as there are rows in the data.
.RE

.nf
.B filename.lines
.fi
.RS .25i

Line segments: specifications for the pattern of line segments which
connect pairs of points.  The file should contain two numbers per
line.  The pair of numbers represents the row numbers of the two
points that should be connected.
.RE

.nf
.B filename.linecolors
.fi
.RS .25i

Line colors: a color for each line in the .lines file.  The file
should contain one color per line.  (It is best if the colors
correspond to the colors used in brushing; see the later section on
resources.)
.RE

.nf
.B filename.nlinkable
.fi
.RS .25i

The number of rows to be linked for brushing and identification.  By
default, nlinkable is equal to the number of rows in the data.  This
feature can be used to link ordinary scatterplots with plots that have
some decorations requiring additional points, such as clustering
trees.
.RE

.nf
.B filename.vgroups
.fi
.RS .25i

Variable groups: an integer for each column in the data.  Each set of
columns that is represented by the same integer will grouped together
for scaling and transformation.  The file is just one long line of
integers.  For example, an input file with four columns could have
a .vgroups file containing the line 1 2 2 3.  The second and third
columns are then grouped together.  The range of their plotting axes
is be the same, and if column 2 is transformed, column 3 is
transformed at the same time.
.RE

.nf
.B filename.rgroups
.fi
.RS .25i

Row groups: an integer for each row in the data.  Each set of
rows that is represented by the same integer will grouped together
for brushing.  In addition, line groups will be derived from the
row groups:  all lines between points in the same row group will
be in the same line group.  The file is just one column, and the
numbers don't need to be in any sort of sequence.
.RE

.nf
.B filename.missing
.fi
.RS .25i

A file identical in structure to filename.dat, where non-zero values
indicate positions with missing (or censored, or otherwise
exceptional) values.  This file represents the pattern of missing
values in the data; it can be examined in a separate XGobi window by
selecting
.I Launch missing data XGobi...
from the
.I Tools
menu.
.RE

.nf
.B filename.imp
.fi
.RS .25i

Multiple imputations of missing values: Each column should have a full
set of imputed values.  The number of rows needs to be identical to
the number of non-zero values in filename.missing, or the number of
missing codes in filename.dat if filename.missing is not provided..
The imputed values should be given in their order in the data column
by column.  For example, if filename.dat looks like this:

.nf
  10 NA 12 -3
  98  0 10  0
  77  3 NA -5
   1  2 NA 10
  NA NA  5 -8
   0  0 10 12
.fi

(six cases, four variables, five missing values), then filename.imp
with two sets of imputed values could look like this:

.nf
  54  37
   3   2
   4   1
  11  10
  13  11 
.fi

If the second column is selected for imputation (Select 
.I Impute missing values
from the
.I Tools
menu), the full data matrix with imputations looks like this:

.nf
  10  2 12 -3
  98  0 10  0
  77  3 10 -5
   1  2 11 10
  37  1  5 -8
   0  0 10 12
.fi
.RE

.nf
.B filename.resources
.fi
.RS .25i

Resources: a set of datafile-specific XGobi resources, which specify
the size of the plotting window and some user-selection option
settings.  The file is in the format of a standard X resource file.
It can be directly edited so that other resources can be specified.
See the later section on resource files for more information.
.RE
.sp
All of the above files can be created outside of XGobi, using an
editor or other UNIX utilities, and several of them (glyphs and
colors, line segments and line color, resources) can be written out
during an XGobi session, in which case they represent the results of
interactions performed during that session.
.SH X11 RESOURCE FILES
X11 resource files are used to control size, color, layout,
labels, and various other windowing features.  The XGobi
resources you are most likely to find useful set the sizes
of the plotting window and the variable selection panel,
the colors available for brushing, the font types and
sizes, and the title appearing in the title bar of the X
window.  Some examples follow:

In this example, the first 5 of the 10 brushing colors are
specified.  These colors should be drawn from the set of
colors available on your workstation or X terminal, which
you can probably determine using the X command 'showrgb.'

.RS .5i
.nf
*brushColor0: white
*brushColor1: red
*brushColor2: green
*brushColor3: skyblue
*brushColor4: yellow
.fi
.RE

Here, the sizes of the plotting window and variable selection panel
are set.  All units are in pixels.  This file was created
interactively during an XGobi session.

.RS .5i
.nf
*XGobi*PlotWindow.height: 120
*XGobi*PlotWindow.width: 490
*XGobi*VarPanel.width: 390
.fi
.RE

These two resources allow the variable widgets to be reduced
in size, which is helpful when there are many variables
in the data.

.RS .5i
.nf
*XGobi*VarWindow.width: 36
*XGobi*VarLabel.width:  36
.fi
.RE

Here, the title and font have been adjusted for making a videotape.
An explanatory title and a large font are used.

.RS .5i
.nf
*XGobi.title:    XGobi: Experimental Laser Data
*XGobi.iconName: XGobi: Laser
*plotFont: -*-lucida-medium-r-normal-sans-20-*-*-*-*-*-*-*
.fi
.RE

Several logical variables controlling XGobi's behavior can be defined
in a resource file.  These are also represented by buttons, most of
which are found on the Option menu accessed from the main panel.
Descriptions can be found in the help window for each button.

.RS .5i
.nf
*showAxes: True
*showPoints: True
*showLines: True
*linkBrush: True
*linkLineBrush: True
*linkIdentify: True
*linkTour: True
*plotSquare: True
*jumpBrush: True
*reshapeBrush: True
*carryVars: True
*glyphType: 6
*glyphSize: 1
*defaultPrintCmd: lpr -Pps1
.fi
.RE
.SH AUTHORS
.nf
Deborah Swayne (dfs@research.att.com)
Dianne Cook (dicook@iastate.edu)
Andreas Buja (andreas@research.att.com)
.fi
.SH CONTACT
Deborah Swayne (dfs@research.att.com)
.SH SEE ALSO
Sample data files are included in the XGobi distribution.

For papers and further pointers, check the following web page:
.nf
http://www.research.att.com/areas/stat/xgobi/
.fi

XGobi supports Remote Procedure Calls (RPCs) as a means for
interprocess communication (IPC). For example, a link among
XGobi, ArcView, and XploRe is currently being supported.
For more details, see:
.nf
http://www.galaxy.gmu.edu/~symanzik/AXX/
.fi

.SH BUGS
Probably too numerous to mention, but you do the authors a favor if
you report the ones you notice.



Changes
September 2001:

  A great many extensions were added to xgvis, and
  a few bugs repaired in xgobi.

August 5 2000:

  Sphering is handled as a transformation instead of being
  buried inside the touring code.

January 3 2000:

  The initial CVS repository on franz, really just a developer's
  version.  The CORBA functionality is working, but still needs
  work, and is not yet documented.  There is a problem with the
  combination of Imakefile/GNUmakefile.

  The menus on each variable are now variable selection menus
  rather than transformation menus.

late November 1999:

  Variable transformations have been moved from the variable
  labels and given their own panel.  A variety of power transformations
  have been replaced by the Box-Cox family of transformations.  
  Transformation has been given a preparatory stage, too, which
  we're calling "domain adjustment:"  ie, ensure that the data
  you want to transform is within the domain of the transformation
  function.  (We're preparing to break transformation into two
  stages, with permutation, sorting and sphering moved to the
  second stage.)

  1D plotting:  David Scott's 1D ASH (Average Shifted Histogram)
  has been added to the textured dotplot as an alternative 1D plot.
  You can now display these plots either horizontally or vertically

  Axes:  The axes will now behave a bit better -- the labels should
  be smart enough not to draw on top of each other.  In addition,
  the Display menu has one more option, for displaying gridlines
  when in the 1D and 2D plots when axes are shown.

  Identify:  no label will be drawn unless you're within a few
  hundred pixels of some point.

  Linked brushing in the presence of both row groups and nlinkable: 
  few of you will have discovered these options, so even fewer of
  you will know that there were tremendous bugs and limitations
  if you wanted to use them together.  It was even worse if you used
  both of those options in conjunction with either subsetting or
  excluding points.  All this should now work.

  CORBA:  If you go so far as to read the source code, you'll notice
  references to CORBA.  This reflects work done by Duncan Temple
  Lang and Ross Ihaka to make xgobi available from omegahat
  (www.omegahat.org); we expect to test and release this work
  in a few weeks.

late June 1999:
  There are more revisions of the parallel coordinates display -- finally
  it can display labels instead of variable indices.  Long labels will
  be truncated to fit.  If you'd like to supply shorter labels, add them
  to the fname.col file like this:  longname|shortname
  
  Erase brushing has been completely redesigned.  There's no longer an
  erase operation on the brushing panel.  Instead, you brush the points
  you'd like to erase with some chosen glyph/color, and then open the
  "Hide or exclude" tool.  Using that tool, you can hide cases (ie,
  don't draw them) or exclude them (ie, don't draw them <and> don't
  include them in calculations for scaling or sphering or projection
  pursuit.)  If that seems too elaborate for a quick-and-dirty erase,
  just brush them in the background color.

early June 1999:
  There are some revisions in the parallel coordinates display --
  there are now actual parallel coordinates drawn.

  There's a new pair of command line arguments:
   -only n/N : read only a sample of n of N lines
   -only a,n : read only from line a to a+n

  The subset panel has been redesigned in the hope that it
  will be easier to use; its functionality has not changed.

  A couple of new transformations have been added; some of
  these will eventually migrate to a new panel entitled
  "Inference."

  The label of the currently identified point can be captured
  using the 'w' key, and then pasted into another window.

  xgvis changes:
  
  New command line arguments:
    -dims n   specify the dimensionality of the embedding space
    -stepsize x  specify the initial stepsize, a value between 0 and 1
  
March 1999:
  The biggest change in this distribution is the addition
  of the -scatmat command line argument, which instructs
  xgobi to display the input data as a scatterplot matrix,
  which can be linked to other xgobis.

This release (November 1998) has the
following changes:

 Probably the biggest change in this distribution is that
 another program, xgvis, has been included.  xgvis is an
 interactive multidimensional scaling program which has
 its own control panel but uses xgobi as its main data
 display.


This release (March 1998) has the following changes:

 XGobi supports Remote Procedure Calls (RPCs) as a means
 forinterprocess communication (IPC). For example, a link among
 XGobi, ArcView, and XploRe is currently being supported.  See
 http://www.galaxy.gmu.edu/~symanzik/AXX/ for more details.

 There's a new mode on the view menu -- moving points.  This
 is not intended as a way to fix outliers; rather, it's useful
 for graph viewing or manipulating maps or other structures
 that are drawn along with data.

 rgroups, lgroups:  a method for grouping rows so that brushing
 one member causes all other members of the same group to be
 brushed at the same time.  The method is extended to lines, too:
 all lines between points in the same group form a lines group.

This release (October, 1997) has the following changes:

  There is a new command line option, -subset n, that allows
  a subset of the data to be specified:  that is, all the data
  is read in, but an intial subset is used for display.

  The main control panel has been redesigned, so that it
  now uses several menus.  Everything on the View menu
  has been there for a while, while much of what's on the
  Tools menu is new:  Smoothing, jittering, the variable
  and case lists, and some methods for working with missing
  data.

  The case profile has been renamed 'parallel coordinates
  plot,' since that's what it is if all profiles are displayed.

  The file I/O has been gathered from here and there on

S/xgobi.S
xgobi <- function(matrx,
                  collab     = dimnames(matrx)[[2]],
                  rowlab     = dimnames(matrx)[[1]],
                  colors     = NULL,
                  glyphs     = NULL,
                  erase      = NULL,
                  lines      = NULL,
                  linecolors = NULL,
                  resources  = NULL,
                  title      = NULL,
                  vgroups    = NULL,
                  std        = "mmx",
                  nlinkable  = NULL,
                  subset     = NULL,
                  display    = NULL)
{
    if (!missing(matrx))
    {
        ### data matrix ###
        x <- eval(matrx)
        if (sum(abs(x[!is.na(x)])) == Inf > 0)
        {
          cat("Sorry, xgobi can't handle Infs\n")
          return()
        }
        dfile <- tempfile("unix")
#       cat(t(x), file = dfile, fill = 80)
        x <- as.matrix(x)
        cat(t(x), file=dfile, sep=c(rep(" ",ncol(x)-1),"\n"))

        args <- ""

        ### Column labels ###
        if (missing(collab))
          collab <- dimnames(x)[[2]]
        else {
          # check data type
          if (!is.vector(collab) || !is.character(collab)) {
            cat("The 'collab' argument needs to be a character vector\n")
            return()
          }
        }
        if (length(collab) > 0)
        {
          colfile <- paste(dfile, ".col", sep="")
          cat(collab, file = colfile, sep="\n")
        }

        ### Row labels ###
        if (missing(rowlab))
          rowlab <- dimnames(x)[[1]]
        else {
          # check data type
          if (!is.vector(rowlab) || !is.character(rowlab)) {
            cat("The 'rowlab' argument needs to be a character vector\n")
            return()
          }
        }
        if (length(rowlab) > 0) {
            rowfile <- paste(dfile, ".row", sep="")
            cat(rowlab, file = rowfile, sep="\n")
        }

        ### Variable groups ###
        if (!missing(vgroups)) {
            # check data type
            if (!is.vector(vgroups) || !is.numeric(vgroups)) {
              cat("The 'vgroups' argument needs to be a numeric vector\n")
              return()
            }

            vgroupfile <- paste(dfile, ".vgroups", sep="")
            cat(vgroups, file = vgroupfile, sep="\n")
        }

        ### Colors ###
        if (!missing(colors)) {
            # check data type
            if (!is.vector(colors) || !is.character(colors)) {
              cat("The 'colors' argument needs to be a character vector\n")
              return()
            }
            colorfile <- paste(dfile, ".colors", sep="")
            cat(colors, file = colorfile, sep="\n")
        }
        ### Glyphs ###
        if (!missing(glyphs))
        {
            # check data type
            if (!is.vector(glyphs) || !is.numeric(glyphs)) {
              cat("The 'glyphs' argument needs to be a numeric vector\n")
              return()
            }
            glyphfile <- paste(dfile, ".glyphs", sep="")
            cat(glyphs, file = glyphfile, sep="\n")
        }
        ### Erase ###
        if (!missing(erase))
        {
            # check data type
            if (!is.vector(erase) || !is.numeric(erase)) {
              cat("The 'erase' argument needs to be a numeric vector\n")
              return()
            }
            erasefile <- paste(dfile, ".erase", sep="")
            cat(erase, file = erasefile, sep="\n")
        }

        ### Connected lines ###
        if (!missing(lines))
        {
            # check data type
            if (!is.matrix(lines) || !is.numeric(lines) || dim(lines)[2] != 2) {                cat("The 'lines' argument must be a numeric 2-column matrix\n")
                return()
            }

            linesfile <- paste(dfile, ".lines", sep="")
            unix(paste("rm -f", linesfile), output=F)
            if (nrow(lines) > 0) {
                for (i in 1:nrow(lines))
                    cat(lines[i,], "\n", file = linesfile, append=T)
            }
            ### Line colors ###
            if (!missing(linecolors)) {
                # check data type
                if (!is.vector(linecolors) || !is.character(linecolors)) {
                   cat("The 'linecolors' argument must be a character vector\n")
                   return()
                }
                linecolorfile <- paste(dfile, ".linecolors", sep="")
                cat(linecolors, file = linecolorfile, sep="\n")
            }
        }
        ### Resources ###
        if (!missing(resources))
        {
            # check data type
            if (!is.vector(resources) || !is.character(resources)) {
                cat("The 'resources' argument must be a character vector\n")
                return()
            }
            resourcefile <- paste(dfile, ".resources", sep="")
            cat(resources, file = resourcefile, sep	="\n")
        }
        ### nlinkable ###
        if (!missing(nlinkable)) {
            # check data type
            nlinkable <- as.integer(nlinkable)
            if (length(nlinkable) > 1) {
                cat("The 'nlinkable' argument must be a scalar integer\n")
                return()
            }
            linkablefile <- paste(dfile, ".nlinkable", sep="")
            cat(nlinkable, "\n", file = linkablefile)
        }
        ### subset ###
        subsetarg <- ""
        if (!missing(subset)) {
            # check data type
            subset <- as.integer(subset)
            if (length(subset) > 1) {
                cat("The 'subset' argument must be a scalar integer\n")
                return()
            }
            else if (subset == 0 || subset > nrow(x)) {
                cat("The 'subset' argument must be >0 and <= nrows\n")
                return()
            }
            subsetarg <- paste(" -subset ", subset, sep="")
            args <- paste(args, subsetarg, sep=" ")
        }

        if (!missing(display)) {
            if (!is.character(display))
                warning("display must be a character string")
            else
                args <- paste("-display", display, args)
        }

        if (!missing(title)) {
            if (!is.character(title)) {
                warning("title must be a character string")
                title <- deparse(substitute(matrx))
            }
        }
        else
            title <- deparse(substitute(matrx))
        args <- paste("-title", paste("'", title, "'", sep=""), args)


# Note to installer:
# Here you will need to specify the path to the xgobi executable
# on your system.
        path <- paste("/usr/local/bin/")

        command <- paste(path, "xgobi", sep="")
        command <- paste(command, args)
        command <- paste(command, dfile, "&")
        cat(command, "\n")

        invisible(unix(command,output.to.S=F))
    }
    else
      cat("Matrix argument required\n")
}
